docs/en/stores/supplemental.md

```diff
@@ -0,0 +1,394 @@
+
+-   `get<S>(path: Path<M, S>): S` takes a `Path` object and returns the value in the current state that that path points to
+-   `at<S extends Path<M, Array<any>>>(path: S, index: number): Path<M, S['value'][0]>` returns a `Path` object that points to the provided `index` in the array at the provided path
+-   `path: StatePaths<M>` is a typesafe way to generate a `Path` object for a given path in the state
+-   `apply(operations: PatchOperation<T>[]): PatchOperation<T>[]` applies the provided operations to the current state
+
+### ImmutableState
+
+Dojo Stores provide an implementation of the MutableState interface that leverages [Immutable](). This implementation may provide better performance if there are frequent, deep updates to the store's state. Performance should be tested and verified before fully committing to this implementation.
+
+> Using Immutable
+
+```ts
+import State from './interfaces';
+import Store from '@dojo/framework/stores/Store';
+import Registry from '@dojo/framework/widget-core/Registry';
+import ImmutableState from '@dojo/framework/stores/state/ImmutableState';
+
+const registry = new Registry();
+const customStore = new ImmutableState<State>();
+const store = new Store<State>({ store: customStore });
+```
+
+## Local Storage
+
+Dojo Stores provides a collection of tools to leverage local storage.
+
+The local storage middleware watches specified paths for changes and stores them locally on disk using the `id` provided to the `collector` and structure as defined by the path.
+
+> Local storage middleware
+
+```ts
+export const myProcess = createProcess(
+	'my-process',
+	[command],
+	collector('my-process', (path) => {
+		return [path('state', 'to', 'save'), path('other', 'state', 'to', 'save')];
+	})
+);
+```
+
+The `load` function hydrates a store from `LocalStorage`
+
+> Hydrating state
+
+```ts
+import { load } from '@dojo/framework/stores/middleware/localStorage';
+import { Store } from '@dojo/framework/stores/Store';
+
+const store = new Store();
+load('my-process', store);
+```
+
+Note that data is serialized for storage and the data is overwritten after each process call. This implementation is not appropriate for non-serializable data (e.g. `Date` and `ArrayBuffer`).
+
+# Subscribing to Store changes
+
+The `Store` has an `onChange(path, callback)` method that takes a path or an array of paths and will call a callback function whenever that state changes.
+
+> main.ts
+
+```ts
+const store = new Store<State>();
+const { path } = store;
+
+store.onChange(path('auth', 'token'), () => {
+	console.log('new login');
+});
+
+store.onChange([path('users', 'current'), path('users', 'list')], () => {
+	// Make sure the current user is in the user list
+});
+```
+
+The `Store` also has an `invalidate` event that fires any time the store changes.
+
+> main.ts
+
+```ts
+store.on('invalidate', () => {
+	// do something when the store's state has been updated.
+});
+```
+
+# JSON Patch and Store operations
```

Comment:
As before I don't think we should mention JSON path in the reference guide.

Reply:
There's some advanced functionality here that I want to cover

1. that stores will automatically create the underlying structure
1. that stores will rarely throw an error
1. that adding and replacing past the end of an array will add to the end of an array

Maybe keep this section and remove the references to JSON Patch